cmd/copy.go

```diff
@@ -876,13 +899,67 @@ func (cca *cookedCopyCmdArgs) processCopyJobPartOrders() (err error) {
 	// depending on the source and destination type, we process the cp command differently
 	// Create enumerator and do enumerating
 	switch cca.fromTo {
-	case common.EFromTo.LocalBlob():
-		fallthrough
-	case common.EFromTo.LocalBlobFS():
-		fallthrough
-	case common.EFromTo.LocalFile():
-		e := copyUploadEnumerator(jobPartOrder)
-		err = e.enumerate(cca)
+	case common.EFromTo.LocalBlob(),
+		common.EFromTo.LocalBlobFS(),
+		common.EFromTo.LocalFile():
+		var traverser resourceTraverser
+		var destTraverser resourceTraverser
+
+		dst := cca.destination
+
+		if cca.destinationSAS != "" {
+			destURL, err := url.Parse(dst)
+
+			if err != nil {
+				return err
+			}
+
+			destURL = copyHandlerUtil{}.appendQueryParamToUrl(destURL, cca.destinationSAS)
+			dst = destURL.String()
+		}
+
+		traverser, err = initResourceTraverser(cca.source, cca.fromTo.From(), nil, nil, &cca.followSymlinks, &cca.listOfFilesLocation, cca.recursive, func() {})
+		if err != nil {
+			return err
+		}
+
+		destTraverser, err = initResourceTraverser(dst, cca.fromTo.To(), &ctx, &cca.credentialInfo, &cca.followSymlinks, nil, cca.recursive, func() {})
+		if err != nil {
+			return err
+		}
+
+		isDir := traverser.isDirectory(false)
+		isDestDir := destTraverser.isDirectory(true)
+
+		if isDir && !cca.recursive {
+			return errors.New("cannot copy from container or directory without --recursive or trailing wildcard (/*)")
+		}
```

Comment:
What if isDir == true and isDestDir==false?  On the face of it, that seems to me like it should be an error...

Reply:
We treat the destination as a directory in this case. This retains old functionality where I could safely do something like so:

```azcopy copy /my/local/directory https://myaccount.blob.core.windows.net/container/virdir```